### PR TITLE
CPM-499: add acceptance tests for Upser Product service API

### PR DESCRIPTION
**Pull request by nmarniesse**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Channel/Locale/back/Test/Acceptance/InMemory/InMemoryGetEditableLocaleCodes.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            $ownedCategoryCodes = \array_merge(
+                $ownedCategoryCodes,
+                $this->editableLocalesPerUserGroupName[$groupName] ?? []
+            );
+        }
+
+        return \array_unique($ownedCategoryCodes);
+    }
+
+    public function addOwnedCategoryCode(string $groupName, string $localeCode): void
```
Comment (nmarniesse) [inline src/Akeneo/Channel/Locale/back/Test/Acceptance/InMemory/InMemoryGetEditableLocaleCodes.php:47]: Usually we use a repository to store the information, and the in memory query use the repository. But in this case the repository is in EE (LocaleAccessRepository) and the methods of this repository don't help us for this query.  
So I decided to do simpler and add this method

File: tests/back/Acceptance/User/InMemoryGroupRepository.php [hunk 2]
```diff
@@ -18,8 +18,8 @@
  */
 class InMemoryGroupRepository implements GroupRepositoryInterface, SaverInterface, ObjectRepository
```
Comment (SamirBoulil) [inline tests/back/Acceptance/User/InMemoryGroupRepository.php:19]: Are you using those in your tests ? I don't think you should rely on those global inMemory implementations but rather keep a copy in your new BC?

Comment (nmarniesse) [reply tests/back/Acceptance/User/InMemoryGroupRepository.php:19]: We discussed about that this yesterday with @ahocquard . And we need to discuss with other people to find a proper way to write tests. To be explain our thoughts shortly:

- copy/paste in memories in our BC is better to isolate the BC, and launch dedicated CI. But with this, the business is duplicated in several BC. For example, if the InMemoryGroupRepository is duplicated in Enrichment, Asset, ... It means all the teams need to know and maintain the business of user/group. If the user/group business changes, all teams need to fix their own in memory implementation!
- another solution can be: the team that owns the user/group BC provides the real implementation (public API) **and** the in memory. This way only one team maintain the user/group business. But in another hand, the tests of the BC are not isolated.

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Feature/upsert.feature [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  Scenario: Create an empty product with unknown user
+    When an unknown user tries to upsert the "foo" product
+    Then there is a violation saying the user is unknown
+
+  Scenario: Create an empty product with empty string identifier
+    When the "julia" user upserts the "" product
+    Then there is a violation with message: The product identifier requires a non empty string
+
+  Scenario: Create a product with text attribute value
+    Given a set text value intent on the "a_text-null-null" attribute with the "test" text value
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Feature/upsert.feature:33]: I think this -null-null a little bit weird in a feature file.
But I saw the function parsing this, and there is no beautiful way to know when you write a_text-toto if toto is a channel or locale, and you way of coding this prevent it.
If you have a good idea to remove the -null items, feel free, if not, TANPI.

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Acceptance/Context/ProductContext.php [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    public function thereIsAMessageSayingTheUserIsUnknown(): void
+    {
+        $this->constraintViolationsContext->thereIsAViolationWithMessage('The "-10" user does not exist');
+    }
+
+    private function upsertProduct(UpsertProductCommand $command): void
+    {
+        $violations = $this->validator->validate($command);
+        $this->constraintViolationsContext->add($violations);
+        // @TODO later: call the handler (we cannot do that now because legacy validation is not in memory)
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Acceptance/Context/ProductContext.php:108]: Can you add a card in the backlog to not forget this plz ?

**Review by pierallard — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Feature/upsert.feature (hunk 3 above)
Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Product/back/Test/Feature/upsert.feature]: it's fixed

**Review by cemorin — APPROVED** [review]

**Pull request merged**